pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py

```diff
@@ -0,0 +1,184 @@
+                max_keys = max(local_max_keys, max_keys)
+                n_app_vertices += 1
+
+        self._n_bits_machine = self._bites_needed(max_machine_vertices)
+        self._n_bits_partition = self._bites_needed(max_partitions)
+        self._n_bits_atoms = self._bites_needed(max_keys)
+        n_bits_vertices = self._bites_needed(n_app_vertices)
+
+        if sum((self._n_bits_machine,
+                self._n_bits_partition,
+                self._n_bits_atoms,
+                n_bits_vertices)) > KEY_SIZE:
+            raise PacmanRouteInfoAllocationException(
+                "Unable to use GlobalZonedRoutingInfoAllocator as it needs "
+                "{} + {} + {} + {} bits".format(
+                    self._n_bits_machine, self._n_bits_partition,
+                    self._n_bits_atoms, n_bits_vertices))
+        return max_partitions
+
+    def _simple_allocate(self):
+        progress = ProgressBar(
+            self._application_graph.n_vertices, "Allocating routing keys")
+        routing_infos = RoutingInfo()
+        by_app_vertex = dict()
+        mask = 0xFFFFFFFF - ((2 ** self._n_bits_atoms) - 1)
+        app_bits = (
+            self._n_bits_atoms + self._n_bits_machine + self._n_bits_partition)
+        app_mask = 0xFFFFFFFF - ((2 ** app_bits) - 1)
+        app_index = 0
+        for app_vertex in progress.over(self._application_graph.vertices):
+            machine_vertices = self._graph_mapper.get_machine_vertices(
+                app_vertex)
+            machine_index = 0
+            for vertex in machine_vertices:
+                partitions = self._machine_graph.\
+                    get_outgoing_edge_partitions_starting_at_vertex(vertex)
+                part_index = 0
+                for partition in partitions:
+                    if partition.traffic_type == EdgeTrafficType.MULTICAST:
+                        key = app_index
+                        key = (key << self._n_bits_machine) | machine_index
+                        key = (key << self._n_bits_partition) | part_index
+                        key = key << self._n_bits_atoms
+                        key_and_mask = BaseKeyAndMask(
+                            base_key=key, mask=mask)
+                        info = PartitionRoutingInfo(
+                            [key_and_mask], partition)
+                        routing_infos.add_partition_info(info)
+                        part_index += 1
+                if part_index > 0:
+                    machine_index += 1
+            if machine_index > 0:
+                app_key = app_index << (self._n_bits_machine +
+                                        self._n_bits_partition +
+                                        self._n_bits_atoms)
+                by_app_vertex[app_vertex] = BaseKeyAndMask(
+                    base_key=app_key, mask=app_mask)
+                app_index += 1
+
+        return routing_infos, by_app_vertex
+
+    def _bites_needed(self, size):
```

Comment:
definte static method here.

Reply:
spelling problem here. if you want to not have it static. named constant for the 2? else, it should go to some helpful functions somewhere.

Reply:
Think you are looking at outdated code...

Regarding the 2, this is the binary base power.  It should be well understood without needing to be named.